README.md
# @amille/dev-scripts

![npm (tag)](https://img.shields.io/npm/v/@amille/dev-scripts/latest)
![npm (tag)](https://img.shields.io/npm/v/@amille/dev-scripts/next)
![CircleCI](https://img.shields.io/circleci/build/github/amille44420/dev-scripts)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

Bunch of webpack scripts ready to use for node and browser react projects.

## Installation

First add `@amille/dev-scripts` into your development dependencies.

```sh
# using npm
npm install --save-dev @amille/dev-scripts

# or using yarn
yarn add -D @amille/dev-scripts
```

You may now move onto the next section related to your needs.

#### Setup for browser projects

You project must follow this setup :

```
| - packages.json
| - /build (output directory)
| - /src (source code)
|   | - index.jsx (bundle entry point)
| - /public (public directory)
|   | - index.html (template file for your bundle html entry)
| - dev-scripts.hooks.js (optional hook file)
```

You may add the following scripts to your owns

```json
{
    "start": "am-scripts start-browser",
    "build": "am-scripts build-browser",
    "build:profile": "yarn build --profile"
}
```

#### Setup for node projects

You project must follow this setup :

```
| - packages.json
| - /build (output directory)
| - /src (source code)
|   | - index.jsx (server entry point)
```

You may add the following scripts to your owns

```json
{
    "start": "am-scripts start-node",
    "build": "am-scripts build-node"
}
```

This configuration has `express` as `peerDependency`.

## Customization

This project may be customize by specifying environment variables.
The scripts will also load environment from env files using `dotenv`.

| name                           | default value | description                                               |
| ------------------------------ | ------------- | --------------------------------------------------------- |
| NODE_ENV                       |               | targeted environment                                      |
| PORT                           | 3000          | Port on which the dev server will listen                  |
| HOST                           | 0.0.0.0       | Host on which the dev server will listen                  |
| IMAGE_INLINE_SIZE_LIMIT        | 10000         | Image size limit for inlining                             |
| HTTPS                          |               | if set to `true`, the dev server will run on HTTPS        |
| APP\_\*                        |               | Environment variables to inject (cf documentation)        |
| DANGEROUSLY_DISABLE_HOST_CHECK |               | if set to `true`, the dev server will disable host checks |

#### Global variables

You may define global variables using environment variables.
Those will be injected with define or interpolate plugins.

In addition of `__DEV__`, all variables which their name start with `APP_` will be use for globals.
The prefix `APP_` will be stripped.

You may use them in your html entry file with the following syntax

```html
<p>my global is %NAME%</p>
```

or in your javascript files

```js
console.log(`my global is ${process.env.NAME}`);
```

## Environment configuration

You may update application settings using environment files.

* `.env`: Default.
* `.env.local`: Local overrides. This file is loaded for all environments except test.
* `.env.development`, `.env.production`: Environment-specific settings.
* `.env.development.local`, `.env.production.local`: Local overrides of environment-specific settings.

Files on the left have more priority than files on the right:

* **`npm run start-*`**: `.env.development.local`, `.env.development`, `.env.local`, `.env`
* **`npm run build-*`**: `.env.production.local`, `.env.production`, `.env.local`, `.env`

These variables will act as the defaults if the machine does not explicitly set them.

### Override

This package comes with many available hooks to override configurations at specific key points.
To setup hooks you simply need to create a file `dev-scripts.hooks.js` at your root directory.

```js
const hooks = require('@amille/dev-scripts/helpers/hooks');

/* for all hooks, "any" can be replaced by "node" or "browser" to run on specific build target */

/* overrideOptions, overrideEnv & overridePaths allows you to override settings */

hooks.any.overrideOptions((options, settings) => {
    // options must be returned
    return options;
});

hooks.any.overrideEnv((env, settings) => {
    // env must be returned
    return env;
});

hooks.any.overridePaths((paths, settings) => {
    // paths must be returned
    return paths;
});

/* overrideSettings allows to override the whole settings object */

hooks.any.overrideSettings(settings => {
    // settings must be returned
    return settings;
});

/* overrideWebpack allows to override the whole webpack config */

hooks.any.overrideWebpack((webpackConfig, settings) => {
    // webpackConfig must be returned
    return webpackConfig;
});

/* setupProxyOnBefore & setupProxyOnAfter are helpful to setup proxies in react web dev servers lifecycle */

hooks.any.setupProxyOnBefore((app, server, settings) => {
    /* ... */
});

hooks.any.setupProxyOnAfter((app, server, settings) => {
    /* ... */
});
```

#### React refresh (experimental)

This package provide an experimental HMR based on `react-refresh`. 
To enable it you must either override the option `withReactRefresh` using hooks.

```js
hooks.browser.overrideOptions(options => {
    options.withReactRefresh = true;

    return options;
});
```

#### Hot reload entry point

The hot reload for browser apps is based on `react-deep-force-updates`.
Setup your entry point as the following example.

**if you use `react-refresh` experimental feature, you don't have to do any of this for your HMR**

```js
import { createElement } from 'react';
// react-deep-force-update will be nullify by null-loader for production
// so we can safely use it for development
// eslint-disable-next-line import/no-extraneous-dependencies
import deepForceUpdate from 'react-deep-force-update';
import { render } from 'react-dom';
// import your app component
import App from './App';

// the query selector must match the one from your html entry point
const container = document.querySelector('#root');

let appInstance = null;

// render core method
const renderApp = () => {
    try {
        // we instance our react element
        const appElement = createElement(App, {});
        // render it and get back the instance
        appInstance = render(appElement, container);
    } catch (error) {
        if (__DEV__) {
            // for development we throw the errors back
            throw error;
        }

        // and print errors in console anyway
        console.error(error);
    }
};

// first rendering
renderApp();

if (module.hot) {
    // hot reload module is here, meaning it is development
    // we listen for all changes in our app
    module.hot.accept('./App', () => {
        if (appInstance && appInstance.updater.isMounted(appInstance)) {
            // Force-update the whole tree, including components that refuse to update
            deepForceUpdate(appInstance);
        }

        renderApp();
    });
}
```


CHANGELOG.md
## [2.1.9](https://github.com/amille44420/dev-scripts/compare/v2.1.8...v2.1.9) (2020-05-11)

## [2.1.8](https://github.com/amille44420/dev-scripts/compare/v2.1.7...v2.1.8) (2020-05-08)

## [2.1.7](https://github.com/amille44420/dev-scripts/compare/v2.1.6...v2.1.7) (2020-05-06)

## [2.1.6](https://github.com/amille44420/dev-scripts/compare/v2.1.5...v2.1.6) (2020-04-28)

## [2.1.5](https://github.com/amille44420/dev-scripts/compare/v2.1.4...v2.1.5) (2020-04-24)

## [2.1.4](https://github.com/amille44420/dev-scripts/compare/v2.1.3...v2.1.4) (2020-04-23)

## [2.1.3](https://github.com/amille44420/dev-scripts/compare/v2.1.2...v2.1.3) (2020-04-22)

## [2.1.2](https://github.com/amille44420/dev-scripts/compare/v2.1.1...v2.1.2) (2020-04-20)

## [2.1.1](https://github.com/amille44420/dev-scripts/compare/v2.1.0...v2.1.1) (2020-04-20)

# [2.1.0](https://github.com/amille44420/dev-scripts/compare/v2.0.17...v2.1.0) (2020-04-13)


### Features

* improve environment configuration ([327c1bb](https://github.com/amille44420/dev-scripts/commit/327c1bbd58986e500f4bc6396cdc8265b29e53c0))

## [2.0.17](https://github.com/amille44420/dev-scripts/compare/v2.0.16...v2.0.17) (2020-04-13)

## [2.0.16](https://github.com/amille44420/dev-scripts/compare/v2.0.15...v2.0.16) (2020-04-08)

## [2.0.15](https://github.com/amille44420/dev-scripts/compare/v2.0.14...v2.0.15) (2020-04-07)

## [2.0.14](https://github.com/amille44420/dev-scripts/compare/v2.0.13...v2.0.14) (2020-04-06)


### Bug Fixes

* fix invalid babel cache identifier ([1ba50c0](https://github.com/amille44420/dev-scripts/commit/1ba50c03d073f407a4dc35db7b6d26babf2745cf))

## [2.0.13](https://github.com/amille44420/dev-scripts/compare/v2.0.12...v2.0.13) (2020-04-02)

## [2.0.12](https://github.com/amille44420/dev-scripts/compare/v2.0.11...v2.0.12) (2020-03-30)

## [2.0.11](https://github.com/amille44420/dev-scripts/compare/v2.0.10...v2.0.11) (2020-03-30)

## [2.0.10](https://github.com/amille44420/dev-scripts/compare/v2.0.9...v2.0.10) (2020-03-26)

## [2.0.9](https://github.com/amille44420/dev-scripts/compare/v2.0.8...v2.0.9) (2020-03-25)

## [2.0.8](https://github.com/amille44420/dev-scripts/compare/v2.0.7...v2.0.8) (2020-03-25)

## [2.0.7](https://github.com/amille44420/dev-scripts/compare/v2.0.6...v2.0.7) (2020-03-24)

## [2.0.6](https://github.com/amille44420/dev-scripts/compare/v2.0.5...v2.0.6) (2020-03-23)

## [2.0.5](https://github.com/amille44420/dev-scripts/compare/v2.0.4...v2.0.5) (2020-03-23)

## [2.0.4](https://github.com/amille44420/dev-scripts/compare/v2.0.3...v2.0.4) (2020-03-16)

## [2.0.3](https://github.com/amille44420/dev-scripts/compare/v2.0.2...v2.0.3) (2020-03-09)

## [2.0.2](https://github.com/amille44420/dev-scripts/compare/v2.0.1...v2.0.2) (2020-03-06)

## [2.0.1](https://github.com/amille44420/dev-scripts/compare/v2.0.0...v2.0.1) (2020-03-02)

# [2.0.0](https://github.com/amille44420/dev-scripts/compare/v1.1.6...v2.0.0) (2020-03-02)


### Bug Fixes

* remove invalid plugins from babel loader for react production build ([7ad69d8](https://github.com/amille44420/dev-scripts/commit/7ad69d80c84194b499c461a2bc57488c69329d09))
* **WebpackPackagePlugin:** fix mistake in checks ([23cc740](https://github.com/amille44420/dev-scripts/commit/23cc7403f0905fdee809dac504ebccec1eeba434))
* fix hooks and start scripts ([d0f6283](https://github.com/amille44420/dev-scripts/commit/d0f6283dfa234c64e1fb01efc40917d9657a4349))
* pass options to getStyleLoaders ([d04eacc](https://github.com/amille44420/dev-scripts/commit/d04eacc9ce5c7da970bb48900b8cf279f1775759))


### Features

* create build directory using recursive option ([558085b](https://github.com/amille44420/dev-scripts/commit/558085b8ea69f8c757c848c9ef622c87fcee8f2e))
* **WebpackPackagePlugin:** customize stringify to get a cleaner package file ([943c4ae](https://github.com/amille44420/dev-scripts/commit/943c4aee65771402db300e15805fffa444dd78b3))
* add custom webpack plugin WebpackPackagePlugin ([f502b62](https://github.com/amille44420/dev-scripts/commit/f502b62e183d90095dfd5115abd19903cf35c166))
* allow experimental usage of react-refresh for HMR ([043ab8d](https://github.com/amille44420/dev-scripts/commit/043ab8d3cdfe747f835234948b8d132fc424c295))
* move every env settings into the option step ([7d7b03c](https://github.com/amille44420/dev-scripts/commit/7d7b03c69a7567b3df789bbbb40d666125387f34))
* provide hook system to have dynamic settings on runtime ([8f9d405](https://github.com/amille44420/dev-scripts/commit/8f9d40577068d83da112ab5d92ac22d18f35a6da))
* stop using default export for node hmr ([fde14a8](https://github.com/amille44420/dev-scripts/commit/fde14a869c3bccabeeb6f6ef5f55aa6b4c5756f6))


### BREAKING CHANGES

* the server entry point must now use module.exports to return the server instance instance of default export
* the whole customization has been updated to hooks

## [1.1.6](https://github.com/amille44420/dev-scripts/compare/v1.1.5...v1.1.6) (2020-02-19)

## [1.1.5](https://github.com/amille44420/dev-scripts/compare/v1.1.4...v1.1.5) (2020-02-12)

## [1.1.4](https://github.com/amille44420/dev-scripts/compare/v1.1.3...v1.1.4) (2020-02-11)

## [1.1.3](https://github.com/amille44420/dev-scripts/compare/v1.1.2...v1.1.3) (2020-02-07)

## [1.1.2](https://github.com/amille44420/dev-scripts/compare/v1.1.1...v1.1.2) (2020-02-05)

## [1.1.1](https://github.com/amille44420/dev-scripts/compare/v1.1.0...v1.1.1) (2020-02-01)

# [1.1.0](https://github.com/amille44420/dev-scripts/compare/v1.0.1...v1.1.0) (2020-01-26)


### Features

* **webpack-node:** add BannerPlugin to webpack configuration ([5063fd9](https://github.com/amille44420/dev-scripts/commit/5063fd9ac3ef07143d18af870ef43046c8019c5d))

## [1.0.1](https://github.com/amille44420/dev-scripts/compare/v1.0.0...v1.0.1) (2020-01-24)


### Bug Fixes

* **webpack-browser:** remove undefined rules ([0e5ddf6](https://github.com/amille44420/dev-scripts/commit/0e5ddf618a5a5d0d08a4e428a8f4d2f0512ec2f8))

# 1.0.0 (2020-01-24)


### Bug Fixes


configs/options.js
const getOptions = () => {
    const env = process.env.NODE_ENV || 'development';
    const isEnvProduction = env === 'production';
    const host = process.env.HOST || '0.0.0.0';
    const port = parseInt(process.env.PORT, 10) || 3000;

    return {
        env,
        // bundle configuration
        envPublicUrl: '/', // for now only support root serving
        imageInlineSizeLimit: parseInt(process.env.IMAGE_INLINE_SIZE_LIMIT || '10000', 10),
        isEnvProductionProfile: isEnvProduction && process.argv.includes('--profile'),
        // style files regexes
        cssRegex: /\.css$/,
        cssModuleRegex: /\.module\.css$/,
        lessRegex: /\.less$/,
        lessModuleRegex: /\.module\.less$/,
        // environments
        isEnvDevelopment: env === 'development',
        isEnvProduction,
        // dev web server config
        host,
        port,
        protocol: process.env.HTTPS === 'true' ? 'https' : 'http',
        disableHostCheck: process.env.DANGEROUSLY_DISABLE_HOST_CHECK === 'true',
        // experimental feature(s)
        withReactRefresh: false,
    };
};

module.exports = getOptions;
